File: backend/app/routers/direct_messages.py
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from ..database import get_db
from ..dependencies.auth import get_current_user
from ..models.user import User
from ..schemas.direct_message import DirectMessageCreate, DirectMessagePage, DirectMessageRead
from ..services.direct_messages import DirectMessageService

router = APIRouter(prefix="/direct-messages", tags=["direct-messages"])


@router.get("/{partner_id}", response_model=DirectMessagePage)
def list_direct_messages(
    partner_id: str,
    limit: int = Query(50, ge=1, le=200),
    before: Optional[datetime] = None,
    me: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> DirectMessagePage:
    messages = DirectMessageService.list_messages(
        db,
        user_id=me.id,
        partner_id=partner_id,
        limit=limit,
        before=before,
    )
    return DirectMessagePage(messages=messages)


@router.post("/{partner_id}", response_model=DirectMessageRead, status_code=201)
def send_direct_message(
    partner_id: str,
    payload: DirectMessageCreate,
    me: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> DirectMessageRead:
    message = DirectMessageService.send_message(
        db,
        sender_id=me.id,
        recipient_id=partner_id,
        content=payload.content,
    )
    return message

File: frontend/src/services/directMessages.ts
import api from '../lib/api';

type ApiDirectMessage = {
  id: string;
  sender_id: string;
  recipient_id: string;
  content: string;
  created_at: string;
};

type ApiDirectMessagePage = {
  messages: ApiDirectMessage[];
};

export type DirectMessage = {
  id: string;
  senderId: string;
  recipientId: string;
  content: string;
  createdAt: string;
};

function normalizeMessage(payload: ApiDirectMessage): DirectMessage {
  return {
    id: payload.id,
    senderId: payload.sender_id,
    recipientId: payload.recipient_id,
    content: payload.content,
    createdAt: payload.created_at,
  };
}

export const directMessagesService = {
  async list(partnerId: string, limit = 50): Promise<DirectMessage[]> {
    const { data } = await api.get<ApiDirectMessagePage>(`/direct-messages/${partnerId}`, {
      params: { limit },
    });
    return data.messages.map(normalizeMessage);
  },
  async send(partnerId: string, content: string): Promise<DirectMessage> {
    const { data } = await api.post<ApiDirectMessage>(`/direct-messages/${partnerId}`, { content });
    return normalizeMessage(data);
  },
};
